### PR TITLE
faster generation of alternate map matches without limitation

### LINKED ISSUE #978: Top K Map Matching Results Remove Hardcoded Uniqueness Limit
In #970 we noted that there were a number of possible paths that can come out of the candidates for the given input points. In #975 we made a change to take care of that but after doing so it was noticed that the bulk of time is spent going through and rejecting thousands of redundant paths. To combat this very long loop that essentially produces hardly any results we put a hardcoded cap in place that basically says dont compute any more than 20 different permutations of candidates.

@ptpt noted that instead we could possibly remove candidates that would produce only redundant paths thereby greatly reducing the number of permutations one would have to visit to find all unique paths for a given input. indeed this approach would gaurantee that you'd only see the unique paths for a given input (i think). his original comment was:

>  After we get the red path (the best one), we can remove candidates which are impossible to be used to construct distinct paths from the red path. For example, the candidate 1 and candidate 2 can be removed. Because their paths to the candidates at next time (3, 4, 5, 6) can't make any difference from the path starting from candidate 0 (via 4TH ST). By removing these "obviously impossible" candidates, we should be able to get the green path in a few iterations or just next iteration.

intro

### PR DESCRIPTION
**Pull request by kevinkreiser**
fixes #978 

@ptpt and I essentially implemented exactly what was discussed in the issue. we basically deleted candidates from the states that had paths that didnt differ from already found results. in doing so we greatly reduce the number of times one has to do the viterbi search algorithm. in fact we've removed the whole notion of checking whether a whole path is unique against a previously found result because change should guarantee that a subsequent viterbi search will find the next unique result.

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -885,8 +885,8 @@ endif
 if DATA_TOOLS
 #some of the tests want tiles of a certain location so lets build those
 test/data/utrecht_tiles: valhalla_build_tiles
-	@echo "Building Utrecht Tiles..." && rm -rf test/data/*utrecht_tiles && ./valhalla_build_tiles --inline-config '{"mjolnir":{"tile_dir":"test/data/tmp_utrecht_tiles","concurrency":1,"logging":{"type":""}}}' test/data/utrecht_netherlands.osm.pbf && mv test/data/tmp_utrecht_tiles test/data/utrecht_tiles
```
Comment (dgearhart) [inline Makefile.am]: was this a mistake? comment out

Comment (kevinkreiser) [reply Makefile.am]: yeah there is other printing stuff in there too. basically the pr isnt ready for review yet sorry. just starting it so that @ptpt and i can collaborate on it

**Comment by kevinkreiser** [comment]
ok did some work on this yesterday and today and fixed a couple things thanks to @ptpt 's advice. the only problem remaining that i can see is with being able to tell which paths are unique and not. there is currently an issue when one of the losing candidates paths is from one node to the same node (different edge though) this ends up being a comparison of no edges and so it looks unique. im trying to think about how to make this check logically consistent..

**Comment by kevinkreiser** [comment]
ok so ive been chasing my tail on this one for a while now and it turns out to be a pretty complicated series of events.

the current status is, everything is much faster wrt topk computations. but the tests are still failing. the reason the test fails is because the first one that @dgearhart added is getting its 2nd most likely alternate as having a uturn and then the 3rd most likely alternate is the path @dgearhart was expecting.

for the longest time i couldnt figure out why. so this morning about 1.5 hours ago i decided i'd hack the code to just return as many unique results as possible without stopping. heres what duanes input looks like:
a

the 1st result with the updated code is as expected:
b

the 2nd result is unfortunate:
c

and the 3rd result is what we were hoping for:
d

For the longest time i couldnt figure out what of all the myriad changes made it so that intuitive second result was happening 3rd instead of 2nd. So i went back to duane branch and moved my changes 1 by 1 to figure out what was doing it leaving the actual removal of candidates until the last thing. it turned out that change that did it was the change to transition_cost_model.cc where we removed the optimization to only add destination candidates who werent routed to before (ie those who didnt have valid predecessor labels, if im understanding the code correctly).

I think what this means is that the test case is actually just not that good! essentially that 2nd result of a uturn is slightly more likely (given the interplay between transition and emission cost) than the 3rd result. just that we didnt see it before because it wasnt even being considered at all (that optimization that i commented out allowed it to be re-computed for `k = 1`).

I'll chat with @ptpt directly to see what he thinks about my hypothesis here.

**Comment by kevinkreiser** [comment]
@ptpt  this is finally ready for review. note that i didnt end up actually making use of removing paths directly (because it was more costly than it was worth). but we could add that if we could find a way to make it faster. in the end we:

1. removed candidates which were entirely redundant
2. changed around some of the topk datastructures so that a few of the common operations were faster
3. removed an optimisation in the viterbi search so that we wouldnt skip computing certain paths. this is needed to allow for removing candidates that are redundant. the test that does 100 random map matches was taking just under 5 seconds on my machine. i'll check what master is and post back. if its slower we can make a setting in viterbi that enables only for requests where `k > 1`

File: src/meili/map_matcher.cc [hunk 2]
```diff
@@ -325,93 +347,242 @@ void MapMatcher::Clear()
...
+      // Cleanup the right hand redundancies
+      for(const auto& r : redundancies) {
+        ts_.RemoveStateId(r);
+        vs_.RemoveStateId(r);
+      }
     }
   }
+}
+
+// Mark any sub paths (pairs of candidates between two states) that are redundant with this one
```
Comment (dgearhart) [inline src/meili/map_matcher.cc]: was this left in in purpose?

File: src/meili/transition_cost_model.cc [hunk 3]
```diff
@@ -135,10 +135,10 @@ TransitionCostModel::UpdateRoute(const StateId& lhs, const StateId& rhs) const
   std::vector<StateId> unreached_stateids;
   unreached_stateids.reserve(right_column.size());
   for (const auto& state : right_column) {
-    if (!vs_.Predecessor(state.stateid()).IsValid()) {
+    //if (!vs_.Predecessor(state.stateid()).IsValid()) {
```
Comment (dgearhart) [inline src/meili/transition_cost_model.cc]: should you remove the if and end bracket?

**Comment by dgearhart** [comment]
high level review looks good to me - I yield to @ptpt for final review

File: valhalla/meili/viterbi_search.h [hunk 4]
```diff
@@ -263,14 +262,13 @@ class NaiveViterbiSearch: public IViterbiSearch
     }
     // remove it from columns
     auto& column = states_[stateid.time()];
+    auto size = column.size();
     const auto it = std::find(column.begin(), column.end(), stateid);
     column.erase(it);
-    // clear current search status to remove possible uses of the removed state ID
-    ClearSearch();
-    return true;
+    return size != column.size();
```
Comment (ptpt) [inline valhalla/meili/viterbi_search.h]: Why is this needed? Isn't it always true?

File: src/meili/topk_search.cc [hunk 5]
```diff
@@ -83,34 +114,32 @@ void EnlargedViterbiSearch::ClonePath(const StateId::Time& time)
...
-      current = origin;
-      last_valid_origin = origin;
-    } // otherwise current is a clone in the other graphs
-  }
+  // Have it clone the current path
+  evss_.back()->ClonePath(path);
+}
 
-  return last_valid_origin;
+StateId TopKSearch::GetOrigin(const StateId& stateid, const StateId& not_found) const
```
Comment (ptpt) [inline src/meili/topk_search.cc]: 👍 for this change!

File: src/meili/topk_search.cc [hunk 6]
```diff
@@ -3,14 +3,26 @@
 namespace valhalla {
 namespace meili {
 
+float EnlargedEmissionCostModel::operator()(const StateId& stateid) {
+  // If this was a removed state bail
+  const auto& original_stateid = evs_.GetOrigin(stateid);
```
Comment (ptpt) [inline src/meili/topk_search.cc]: what if `original_stateid` is invalid? i.e. stateid is origin.

File: valhalla/meili/topk_search.h [hunk 7]
```diff
@@ -32,24 +36,30 @@ class EnlargedTransitionCostModel
...
         claim_stateid_(claim_stateid),
         original_emission_cost_model_(vs.emission_cost_model()),
         original_transition_cost_model_(vs.transition_cost_model()),
         origin_(),
         clone_(),
         clone_start_time_(kInvalidTime),
-        clone_end_time_(kInvalidTime)
+        clone_end_time_(kInvalidTime),
+        initial_origins_(initial_origins),
+        removed_origins_(removed_origins)
```
Comment (ptpt) [inline valhalla/meili/topk_search.h]: `removed_stateids` might be better because it allows removing clones too.

File: src/meili/topk_search.cc (hunk 6 above)
Comment (kevinkreiser) [reply src/meili/topk_search.cc]: if its invalid `IsRemoved` will return false, we will then search for the unordinal id in the cache, if its not found, the invalid original id gets sent to `calculate_cost` which will see its invalid and not use it

File: valhalla/meili/topk_search.h (hunk 7 above)
Comment (kevinkreiser) [reply valhalla/meili/topk_search.h]: only original ids are ever removed because those are the ones we use in `RemoveRendundancies` to check for redundant paths. these are candidates that are found within the `StateContainer`.

we also use this trick to remove clones by checking above for the clones original id in the enlarged emission cost model. this way we dont have to store clones at all, we simply look what the original id was and if the redundancy check marked it then the emission cost is `-1` and the clones path will be removed

File: valhalla/meili/viterbi_search.h (hunk 4 above)
Comment (kevinkreiser) [reply valhalla/meili/viterbi_search.h]: i had this in here because i was debugging and trying to make sure that what was in the `added_states_` was also in the `states_` i think we've proven that that is ok so we can remove this.

File: src/meili/topk_search.cc (hunk 6 above)
Comment (ptpt) [reply src/meili/topk_search.cc]: It won't return the cost for `stateid` if `stateid` is an origin, will it? if it's an origin, then `original_stateid` will be invalid, and then we get -1 from `calculate_cost`.

Comment (kevinkreiser) [reply src/meili/topk_search.cc]: if `stateid` is an origin, then if its not marked as redundant it will look it up in the cache. if its not in the cache it will call `calculate_cost`, see the origin is invalid and then use the model to get the cost

Comment (ptpt) [reply src/meili/topk_search.cc]: I get it. You just check if the original stateid is removed or not.

Comment (kevinkreiser) [reply src/meili/topk_search.cc]: yep!
